com.unity.netcode.gameobjects/Runtime/Serialization/FastBufferReaderExtensions.cs

```diff
@@ -0,0 +1,286 @@
+                value = go;
+                return;
+            }
+
+            if (type == typeof(NetworkObject))
+            {
+                reader.ReadValueSafe(out NetworkObject no);
+                value = no;
+                return;
+            }
+
+            if (typeof(NetworkBehaviour).IsAssignableFrom(type))
+            {
+                reader.ReadValueSafe(out NetworkBehaviour nb);
+                value = nb;
+                return;
+            }
+            /*if (value is INetworkSerializable)
+            {
+                //TODO ((INetworkSerializable)value).NetworkSerialize(new NetworkSerializer(this));
+                return;
+            }*/
+
+            throw new ArgumentException($"{nameof(FastBufferReader)} cannot read type {type.Name} - it does not implement {nameof(INetworkSerializable)}");
+        }
+
+        /// <summary>
+        /// Read an INetworkSerializable
+        /// </summary>
+        /// <param name="value">INetworkSerializable instance</param>
+        /// <typeparam name="T"></typeparam>
+        /// <exception cref="NotImplementedException"></exception>
+        public static void ReadNetworkSerializable<T>(this ref FastBufferReader reader, out T value) where T : INetworkSerializable
+        {
+            throw new NotImplementedException();
+        }
+
+        /// <summary>
+        /// Read a GameObject
+        /// </summary>
+        /// <param name="value">value to read</param>
+        public static void ReadValue(this ref FastBufferReader reader, out GameObject value)
+        {
+            reader.ReadValue(out ulong networkObjectId);
+
+            if (NetworkManager.Singleton.SpawnManager.SpawnedObjects.TryGetValue(networkObjectId, out NetworkObject networkObject))
+            {
+                value = networkObject.gameObject;
+                return;
+            }
+
+            if (NetworkLog.CurrentLogLevel <= LogLevel.Normal)
+            {
+                NetworkLog.LogWarning($"{nameof(FastBufferReader)} cannot find the {nameof(GameObject)} sent in the {nameof(NetworkSpawnManager.SpawnedObjects)} list, it may have been destroyed. {nameof(networkObjectId)}: {networkObjectId}");
+            }
+
+            value = null;
+        }
+
+        /// <summary>
+        /// Read a GameObject
+        ///
+        /// "Safe" version - automatically performs bounds checking. Less efficient than bounds checking
+        /// for multiple reads at once by calling TryBeginRead.
+        /// </summary>
+        /// <param name="value">value to read</param>
+        public static void ReadValueSafe(this ref FastBufferReader reader, out GameObject value)
+        {
+            reader.ReadValueSafe(out ulong networkObjectId);
+
+            if (NetworkManager.Singleton.SpawnManager.SpawnedObjects.TryGetValue(networkObjectId, out NetworkObject networkObject))
+            {
+                value = networkObject.gameObject;
+                return;
+            }
```

Comment:
we intentionally killed GameObject, NetworkObject & NetworkBehaviour serialization from built-in serializers. we wanted to kill NetworkManager/SpawnManager dependency in a primitive impl like this to keep it isolated and dependency-free (as much as we can).
also we were planning to offer more focused NetworkObject/NetworkBehaviour wrappers which would implement INetworkSerializable interface: `struct NetworkObjectReference : INetworkSerializable` etc.

Reply:
I agree with that approach... I'll look into doing that in the INetworkMessage branch when I refactor INetworkSerializable.